File: 3852-path-existence-queries-in-a-graph-ii/path-existence-queries-in-a-graph-ii.py
class Solution:
    def pathExistenceQueries(self, n: int, nums: List[int], maxDiff: int, queries: List[List[int]]) -> List[int]:
        pos = sorted(range(n), key=lambda x: nums[x])
        
        sorted_pos = [0] * n
        for idx, orig in enumerate(pos):
            sorted_pos[orig] = idx
            
        LOG = 18
        st = [[0] * LOG for _ in range(n)]
        
        right = 0
        for i in range(n):
            while right < n and nums[pos[right]] - nums[pos[i]] <= maxDiff:
                right += 1
            st[i][0] = right - 1
            
        for j in range(1, LOG):
            for i in range(n):
                st[i][j] = st[st[i][j-1]][j-1]
                
        ans = []
        
        for u_orig, v_orig in queries:
            u = sorted_pos[u_orig]
            v = sorted_pos[v_orig]
            
            if u > v:
                u, v = v, u
                
            if u == v:
                ans.append(0)
                continue
                
            curr = u
            steps = 0
            
            for j in range(LOG - 1, -1, -1):
                if st[curr][j] < v:
                    curr = st[curr][j]
                    steps += (1 << j)
                    
            if st[curr][0] >= v:
                ans.append(steps + 1)
            else:
                ans.append(-1)
                
        return ans

File: 3852-path-existence-queries-in-a-graph-ii/path-existence-queries-in-a-graph-ii.java
import java.util.Arrays;

class Solution {
    public int[] pathExistenceQueries(int n, int[] nums, int maxDiff, int[][] queries) {
        Integer[] pos = new Integer[n];
        for (int i = 0; i < n; i++) pos[i] = i;
        Arrays.sort(pos, (a, b) -> Integer.compare(nums[a], nums[b]));
        
        int[] sortedPos = new int[n];
        for (int i = 0; i < n; i++) {
            sortedPos[pos[i]] = i;
        }
        
        int LOG = 18; 
        int[][] st = new int[n][LOG];
        
        int right = 0;
        for (int i = 0; i < n; i++) {
            while (right < n && nums[pos[right]] - nums[pos[i]] <= maxDiff) {
                right++;
            }
            st[i][0] = right - 1; 
        }
        
        for (int j = 1; j < LOG; j++) {
            for (int i = 0; i < n; i++) {
                st[i][j] = st[st[i][j - 1]][j - 1];
            }
        }
        
        int[] ans = new int[queries.length];
        
        for (int i = 0; i < queries.length; i++) {
            int u = sortedPos[queries[i][0]];
            int v = sortedPos[queries[i][1]];
            
            if (u > v) {
                int temp = u; u = v; v = temp;
            }
            
            if (u == v) {
                ans[i] = 0;
                continue;
            }
            
            int curr = u;
            int steps = 0;
            
            for (int j = LOG - 1; j >= 0; j--) {
                if (st[curr][j] < v) {
                    curr = st[curr][j];
                    steps += (1 << j);
                }
            }
            
            if (st[curr][0] >= v) {
                ans[i] = steps + 1;
            } else {
                ans[i] = -1; 
            }
        }
        
        return ans;
    }
}